![](https://github.com/JBerthelier/ParasiTE/blob/master/logo.png)

# Identification of TE-genes isoforms events

In brief, ParasiTE is a new tool that aim to: 

1) identify TEs located in exonic or intronic region of genes
2) detect TEs that are part of gene transcripts and the ones that contribute to alternative isoforms of genes (altTE-Gi).

ParasiTE detect candidates for TE-AS and TE-ATP events as illustrate below:

![](https://github.com/JBerthelier/ParasiTE/blob/master/ParasiTE_altTE-Gi_illustration.png)


# Main steps of ParasiTE

ParasiTE is globally composed of five steps:

1) Remove of transcripts of gene-like TEs (transcript of active TEs which are not involved in gene transcripts)
2) Discrimination of intragenic and intergenic TEs
3) Discrimination of intronic and exonic TEs
4) Detection of TE-gene isoforms (TE-Gi) events
5) Detection of alternative TE-gene isoforms (altTE-Gi) events

![](https://github.com/JBerthelier/ParasiTE/blob/master/ParasiTE_steps_illustration.png)

# Prerequise to run ParasiTE in linux

1) R (version 3.6.1 was used) https://www.r-project.org/

2) R libraries: 
- optparse https://cran.r-project.org/web/packages/optparse/readme/README.html
- stringr https://cran.r-project.org/web/packages/stringr/readme/README.html
- data.table https://www.rdocumentation.org/packages/data.table/versions/1.14.2
- dplyr https://www.r-project.org/nosvn/pandoc/dplyr.html
- splitstackshape https://www.r-project.org/nosvn/pandoc/splitstackshape.html
- tidyr https://www.tidyverse.org/blog/2019/09/tidyr-1-0-0/
 
3) bedtools  https://bedtools.readthedocs.io/en/latest/index.html

# Input data

Three inputs are required and one is optional

1) TE annotation in .gff/gtf 
2) gene annotation in .gff/.gtf
3) de novo transcriptome annotation in .gtf file generated by Stringtie2 (https://ccb.jhu.edu/software/stringtie/)
or a gff/gtf file that have the same structure than files generated by Stringtie2 (see below).
4) [OPTIONAL] a gene-like transcript of TE annotation (transcript of active TEs) such as the one of Panda et al. 2020 https://academic.oup.com/plcell/article/32/9/2687/6115671?login=true for A. thaliana

# Outputs

ParasiTE wanted outputs are:

- Annotation of intergenic and intragenic TEs.
- Among Intragenic TEs, the annotation of intragenic (intronic and exonic) TEs.
- List of TE-genes candidates
- List of altTE-gene candidates


# Run ParasiTE (please read until the end)

**The basic command is:**

`Rscript /patway/to/ParasiTE_v1.r -T /Pathway/to/TE_annotation.gff3 -G /pathway/to/gene_annotation.gtf 
-R /pathway/to/transcripts_annotation.gtf -L /pathway/to/gene-like_TE_annotation.gff3 -P {mode}`

**ParasiTE was built to work with Stringtie2 de novo transcriptome (but can use custom transcriptomes).** 

Indeed, in most reference annotation altTE-G transcripts are not resolved by common assemblers (Cufflinks etc ...). 

We choose Stringtie2 https://ccb.jhu.edu/software/stringtie/, because it allows to identify chimeric transcript such as altTE-Gi.

Stringtie2 help to identify potential new isoform that have not been identificate in the reference genome annotation. 

Moreover. Stringtie2 supports short reads (eg. Illumina) or long reads (eg. PacBio or Oxford Nanopore).

**For the -P mode**

1) Transcriptomes obtained with Stringtie2 from long reads alignement

- If the transcriptome was obtained with Stringtie2 with the -L mode you must use `-P SL`   

- If the transcriptome was obtained with Stringtie2 with the -R mode you must use `-P SR`  

2) Transcriptomes obtained with Stringtie2 from short reads alignement
 you must use `-P SL`  

3) Transcriptomes obtained with Stringtie2 'merged'
you must use `-P SM`   

4) Custom transciptomes
The custom transcriptome must follows the stringtie2 format as showed bellow and be run with `-P SA`  

|seqname|source|feature   |start |end  |score|strand|frame|attribute                                                  |
|------:|-----:|---------:|-----:|----:|----:|-----:|----:|----------------------------------------------------------:|
|1      |Ref   |transcript|6788  |9130 |1000 | +    | .   |gene_id "ID.1"; transcript_id "ID.1.1";                    |
|1      |Ref   |exon      |6788  |7069 |1000 | .    | .   |gene_id "ID.1"; transcript_id "ID.1.1"; exon_number_id "1";|
|1      |Ref   |exon      |8571  |9130 |1000 | .    | .   |gene_id "ID.1"; transcript_id "ID.1.1"; exon_number_id "2";|
|1      |Ref   |transcript|3631  |5899 |1000 | +    | .   |gene_id "ID.1"; transcript_id "ID.1.2";                    |
|1      |Ref   |exon      |6788  |7069 |1000 | .    | .   |gene_id "ID.1"; transcript_id "ID.1.2"; exon_number_id "1";|
|1      |Ref   |exon      |7157  |7450 |1000 | .    | .   |gene_id "ID.1"; transcript_id "ID.1.2"; exon_number_id "2";|
|1      |Ref   |exon      |8571  |8737 |1000 | .    | .   |gene_id "ID.1"; transcript_id "ID.1.2"; exon_number_id "3";|
|1      |Ref   |transcript|3631  |5899 |1000 | +    | .   |gene_id "ID.2"; transcript_id "ID.2.1";                    |
|1      |Ref   |exon      |6788  |7069 |1000 | .    | .   |gene_id "ID.2"; transcript_id "ID.2.1"; exon_number_id "1";|
etc...

**Be careful** 
-For now, ParasiTE only works with a "seqname" having a **number** has showed in table (no characters allowed, in example "1" corresponds to "chromosome 1").

"Attribute" must followed the correct stringtie2 format as showed: 

-for transcript:  gene_id "ID.1"; transcript_id "ID.1.1";
-for exon: gene_id "ID.1"; transcript_id "ID.1.1"; exon_number_id "1";

otherwise ParasiTE is not able to works

# Parasite help

`Rscript /patway/to/ParasiTE_v1.r -h`

![](https://github.com/JBerthelier/ParasiTE/blob/master/Help_illustration.PNG)



